R/MsBackendCached.R

```diff
@@ -0,0 +1,616 @@
+                                  collapse = ", "), " not available.",
+             call. = FALSE)
+    if (length(res))
+        as(res, "DataFrame")
+    else NULL
+}
+
+#' @rdname MsBackendCached
+setMethod(
+    "spectraData", "MsBackendCached",
+    function(object, columns = spectraVariables(object)) {
+        .spectra_data(object, columns = columns)
+    })
+
+#' @rdname MsBackendCached
+setReplaceMethod("spectraData", "MsBackendCached",function(object, value) {
+    stop(class(object)[1], " does not support replacing the full spectra data.")
+})
+
+#' @rdname MsBackendCached
+setMethod("[", "MsBackendCached", function(x, i, j, ..., drop = FALSE) {
+    if (missing(i))
+        return(x)
+    slot(x, "localData", check = FALSE) <- x@localData[i, , drop = FALSE]
+    x@nspectra <- nrow(x@localData)
+    x
+})
+
+#' @rdname MsBackendCached
+setMethod("$", "MsBackendCached", function(x, name) {
+    if (!any(spectraVariables(x) == name))
+        stop("Spectra variable '", name, "' not available.")
+    spectraData(x, name)[, 1]
+})
+
+#' @rdname MsBackendCached
+setReplaceMethod("$", "MsBackendCached", function(x, name, value) {
+    if (name %in% c("mz", "intensity"))
+        stop("Replacing m/z and intensity values is not supported.",
+             call. = FALSE)
+    if (length(value) == 1)
+        value <- rep(value, length(x))
+    if (length(value) != length(x))
+        stop("value has to be either of length 1 or length equal to the ",
+             "number of spectra")
+    if (length(x@localData)) {
+        cn <- colnames(x@localData) == name
+        if (any(cn))
+            x@localData[, cn] <- value
+        else {
+            cn <- colnames(x@localData)
+            x@localData <- cbind(x@localData, value)
+            colnames(x@localData) <- c(cn, name)
+        }
+    } else {
+        x@localData <- data.frame(value)
+        colnames(x@localData) <- name
+    }
+    validObject(x)
+    x
+})
+
+#' @rdname MsBackendCached
+setMethod(
+    "selectSpectraVariables", "MsBackendCached",
+    function(object, spectraVariables = spectraVariables(object)) {
+        if (any(!spectraVariables %in% spectraVariables(object)))
+            stop("spectra variable(s) ",
+                 paste(spectraVariables[!spectraVariables %in%
+                                        spectraVariables(object)],
+                       collapse = ", "), " not available")
+        object@spectraVariables <- intersect(object@spectraVariables,
+                                             spectraVariables)
+        object@localData <- object@localData[, colnames(object@localData) %in%
+                                               spectraVariables, drop = FALSE]
+        validObject(object)
+        object
+    })
+
+#' @rdname MsBackendCached
+setMethod("show", "MsBackendCached", function(object) {
+    n <- object@nspectra
+    cat(class(object), "with", n, "spectra\n")
+    if (n) {
+        idx <- union(1:min(6, n), max(1, n-5):n)
```

Comment:
`union` is just `unique(c(x, y))`
```suggestion
        idx <- unique(c(1L:min(6L, n), max(1L, n-5L):n))
```